src/libraries/Common/src/System/HexConverter.cs

```diff
@@ -149,5 +157,95 @@ public static char ToCharLower(int value)
 
             return (char)value;
         }
+
+        public static bool TryDecodeFromUtf16(ReadOnlySpan<char> chars, Span<byte> bytes)
+        {
+            return TryDecodeFromUtf16(chars, bytes, out _);
+        }
+
+        public static bool TryDecodeFromUtf16(ReadOnlySpan<char> chars, Span<byte> bytes, out int charsProcessed)
+        {
+            Debug.Assert(chars.Length % 2 == 0, "Un-even number of characters provided");
+            Debug.Assert(chars.Length / 2 == bytes.Length, "Target buffer not right-sized for provided characters");
+
+            int i = 0;
+            int j = 0;
+            int byteLo = 0;
+            int byteHi = 0;
+            while (j < bytes.Length)
+            {
+                byteLo = FromChar(chars[i + 1]);
+                byteHi = FromChar(chars[i]);
+
+                // byteHi hasn't been shifted to the high half yet, so the only way the bitwise or produces this pattern
+                // is if either byteHi or byteLo was not a hex character.
+                if ((byteLo | byteHi) == 0xFF)
+                    break;
+
+                bytes[j++] = (byte)((byteHi << 4) | byteLo);
+                i+=2;
+            }
+
+            if (byteLo == 0xFF)
+                i++;
+
+            charsProcessed = i;
+            return (byteLo | byteHi) != 0xFF;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int FromChar(int c)
+        {
+            ReadOnlySpan<byte> charToHexLookup = CharToHexLookup;
+            return c >= charToHexLookup.Length ? 0xFF : charToHexLookup[c];
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int FromUpperChar(int c)
+        {
+            ReadOnlySpan<byte> charToHexLookup = CharToHexLookup;
+            return c > 71 ? 0xFF : charToHexLookup[c];
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static int FromLowerChar(int c)
+        {
+            if ('0' <= c && c <= '9')
+                return (byte)(c - '0');
```

Comment:
BTW: The whole method could be written as 
```c#
[MethodImpl(MethodImplOptions.AggressiveInlining)]
public static int FromLowerChar(int c)
{
    if ((uint)(c - '0') <= '9' - '0')
        return c - '0';

    if ((uint)(c - 'a') <= 'f' - 'a')
        return c - 'a' + 10;

    return 0xFF;
}
```
I assume the JIT will evaluate `c - '0'` only once. For `c - 'a'` I'm not sure, because of the addition of 10.
Anyway less branchy code.

Reply:
Good tip! Done de2085b